File: Dart/merge-two-2D-arrays-by-sum-values.dart
List<List<int>> mergeArrays(List<List<int>> nums1, List<List<int>> nums2) {
  Map<int, int> map = {}; // Create a map to store ID and their summed values

  // Iterate through the first list and store values in the map
  for (var pair in nums1) {
    int id = pair[0];
    int value = pair[1];
    map[id] = (map[id] ?? 0) + value; // Add value if exists, else set new
  }

  // Iterate through the second list and add values to the map
  for (var pair in nums2) {
    int id = pair[0];
    int value = pair[1];
    map[id] = (map[id] ?? 0) + value; // Add value if exists, else set new
  }

  // Convert map entries to a list of lists
  List<List<int>> result = map.entries.map((e) => [e.key, e.value]).toList();

  // Sort the result based on the id in ascending order
  result.sort((a, b) => a[0].compareTo(b[0]));

  return result;
}

// Example Usage
void main() {
  print(
    mergeArrays(
      [
        [1, 2],
        [2, 3],
        [4, 5]
      ],
      [
        [1, 4],
        [3, 2],
        [4, 1]
      ],
    ),
  );
  // Output: [[1,6],[2,3],[3,2],[4,6]]

  print(
    mergeArrays(
      [
        [2, 4],
        [3, 6],
        [5, 5]
      ],
      [
        [1, 3],
        [4, 3]
      ],
    ),
  );
  // Output: [[1,3],[2,4],[3,6],[4,3],[5,5]]
}

File: Javascript/merge-two-2D-arrays-by-sum-values.js
function mergeArrays(nums1, nums2) {
    let map = new Map(); // Create a map to store id-value pairs

    // Process the first array
    for (let [id, val] of nums1) {
        map.set(id, (map.get(id) || 0) + val); // If id exists, add to its value; otherwise, set it
    }

    // Process the second array
    for (let [id, val] of nums2) {
        map.set(id, (map.get(id) || 0) + val); // Sum values if id exists
    }

    // Convert the map to an array and sort by id
    return [...map].sort((a, b) => a[0] - b[0]);
}

// Example Usage
console.log(mergeArrays([[1,2],[2,3],[4,5]], [[1,4],[3,2],[4,1]]));
// Output: [[1,6],[2,3],[3,2],[4,6]]

console.log(mergeArrays([[2,4],[3,6],[5,5]], [[1,3],[4,3]]));
// Output: [[1,3],[2,4],[3,6],[4,3],[5,5]]
